### PR TITLE
add report on job & user level, add claimed failed tests in reports

### PR DESCRIPTION
**Pull request by mguillem**
and API to get number of unclaimed failed tests for a job

Note that with this change, unclaimed builds are not shown anymore in the existing report.

### CODE REVIEW

**Comment by Greybird** [comment]
Thanks for the PR, I'll try to take a look in the next days!

File: src/main/java/hudson/plugins/claim/AbstractAssignedClaimsReport.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+		            }
+				}
+            }
+        }
+
+        return claimedTestFailures;
+    }
+    
+    protected boolean isDisplayed(AbstractClaimBuildAction<?> claimAction) {
+    	return claimAction != null && claimAction.isClaimed();
```
Comment (Greybird) [inline src/main/java/hudson/plugins/claim/AbstractAssignedClaimsReport.java:120]: Could this claimed-only by an option in the report ? I'm a little bit reluctant to remove a feature from existing uses, even if I admit that this might not be used at all). 
Would a checkbox with a default of not displaying unclaimed items fit ?

File: src/main/java/hudson/plugins/claim/AbstractAssignedClaimsReport.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	protected List<Job> getJobs() {
+		return Jenkins.getInstance().getAllItems(Job.class);
+	}
+
+    public List<TestResult> getTestFailuresWithClaim() {
+        List<TestResult> claimedTestFailures = new ArrayList<>();
+        for (Job job : getJobs()) {
+            Run lb = job.getLastCompletedBuild();
+            if (lb != null) {
+				AbstractTestResultAction testResultAction = lb.getAction(AbstractTestResultAction.class);
```
Comment (Greybird) [inline src/main/java/hudson/plugins/claim/AbstractAssignedClaimsReport.java:103]: Could you please fix the formatting, and make sure you use spaces and not tabs in files ?

File: src/main/java/hudson/plugins/claim/UserClaimsAction.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	
+	@Override
+	public ModelObject getOwner() {
+		return targetUser;
+	}
+
+
+	@Override
+	protected boolean isDisplayed(AbstractClaimBuildAction<?> claimAction) {
+		return super.isDisplayed(claimAction) && claimAction.getClaimedBy().equals(targetUser.getId());
```
Comment (Greybird) [inline src/main/java/hudson/plugins/claim/UserClaimsAction.java:46]: I was able to get empty reports for a user with actual claims under an instance run by `hpi:run`, with no security configured.
I suspect this check is the culprit, and should probably compare actual `User` objects, taken from `User.get(...)`
I believe at some point, I'll have to take a look at the actual data stored as claimed and assigned data.

File: src/main/java/hudson/plugins/claim/UserClaimsAction.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+
+	@Override
+	public String getDisplayName() {
+		User currentUser = User.current();
+		if (currentUser != null && Objects.equals(targetUser.getId(), currentUser.getId())) {
+			return "My claims";
```
Comment (Greybird) [inline src/main/java/hudson/plugins/claim/UserClaimsAction.java:54]: Both these strings should be made part of Messages class to allow localization.

**Review by Greybird — CHANGES_REQUESTED** [review]

File: src/main/java/hudson/plugins/claim/AbstractAssignedClaimsReport.java (hunk 1 above)
Comment (mguillem) [reply src/main/java/hudson/plugins/claim/AbstractAssignedClaimsReport.java:120]: If I correctly remember, the report currently displays builds with ClaimAction, no matter if they are broken or not and no matter if they are assigned or not. Using the same logic for unit tests with ClaimTestAction doesn't make sense to me as this would mean huge number of (hopefully green) unit tests.
I couldn't find a usefull logic to keep the "old way" and to allow displaying of the interesting Claim(Test)Action: the ones for broken builds & tests. If you have a better idea, I can try to help.

File: src/main/java/hudson/plugins/claim/AbstractAssignedClaimsReport.java (hunk 2 above)
Comment (mguillem) [reply src/main/java/hudson/plugins/claim/AbstractAssignedClaimsReport.java]: Done

File: src/main/java/hudson/plugins/claim/UserClaimsAction.java (hunk 4 above)
Comment (mguillem) [reply src/main/java/hudson/plugins/claim/UserClaimsAction.java]: Done

File: src/main/java/hudson/plugins/claim/UserClaimsAction.java (hunk 3 above)
Comment (mguillem) [reply src/main/java/hudson/plugins/claim/UserClaimsAction.java]: Can you provide some details on how to reproduce? We have a custom build with these changes running since more than two weeks on one server and we haven't noticed any problem until now.

Comment (Greybird) [reply src/main/java/hudson/plugins/claim/UserClaimsAction.java]: I just ran under hpi:run, with no auth or user creation control activated. 
I created another user than the admin one, 
and using the admin one, assigned this user a test
Then I went to the People tab, click on the user, then on his Claims report
No tests were shown.

I believe this is a corner case, and that MOST of the time, this is working under standard scenarios, but I would prefer that Users are compared in a consistent way in the plugin.

Comment (mguillem) [reply src/main/java/hudson/plugins/claim/UserClaimsAction.java]: Can you detail a bit more what you did? 
When i start with `hpi:run`, I have no user at all (neither admin nor anymone else). The only way I've seen to create a user is to use the script console and to execute something like `User.get("foo25", true, [:])` but in this case I can't assign anything to this user, probably because there is still no current user.

File: src/main/java/hudson/plugins/claim/AbstractAssignedClaimsReport.java (hunk 1 above)
Comment (Greybird) [reply src/main/java/hudson/plugins/claim/AbstractAssignedClaimsReport.java]: Ok that's fair. If this is an issue to some users, we could reintroduce the feature only for builds.
Anoway it's not like if these screens would allow to claim unclaimed build/tests...

File: src/main/java/hudson/plugins/claim/UserClaimsAction.java (hunk 3 above)
Comment (Greybird) [reply src/main/java/hudson/plugins/claim/UserClaimsAction.java]: you might be right, I might have a specific configuration that gets reused when running hpi:run as I also have some jobs that I preconfigured.
I think the key might be to configure jenkins security. I have these settings on this instance, could you check  if it would allow you to get the same behavior as I have?
image

Comment (Greybird) [reply src/main/java/hudson/plugins/claim/UserClaimsAction.java]: I managed to debug the issue.

Under these settings, I am able to register a new user, and I make sure the id has uppercase characters by entering every data in upper case.
The user will then have an uppercased id in Jenkins memory, and be serialized in the user database with this uppercased id.
Claiming a test then registers the same uppercased id as the claiming user.

Restarting jenkins will have it reload its user database, and under the same settings, it will use the `CaseInsensitive` `IdStrategy`, which will lower-case the ids on load to make future comparisons faster.
At this point, using the provided code, comparing the users will lead to a case-sensitive comparison which will obviously consider both ids different.

I just pushed a fix.

File: src/test/java/hudson/plugins/claim/UserClaimsReportTest.java [hunk 5]
```diff
@@ -73,7 +74,26 @@ public void userAssignedFailedJobsShouldBeVisibleInUserClaimReport() throws Exce
...
+    }
+
+    @Test
+    public void userAssignedFailedJobsUnderADifferentIdForSameUserShouldBeVisibleInUserClaimReport() throws Exception {
+        long timestamp = System.currentTimeMillis();
+
+        FreeStyleProject job1 = createFailingJobWithName("job1-" + timestamp);
+
+        User userA = User.get("userA-" + timestamp, true, Collections.EMPTY_MAP);
+        assert !userA.getId().equals(userA.getId().toLowerCase()); // ensure id of user is not equal to its lowercase version
```
Comment (mguillem) [inline src/test/java/hudson/plugins/claim/UserClaimsReportTest.java:86]: `assert` is only enabled when the -ea flag is activated, furthermore it doesn't give any helpful message in case of failure. Using some `assertEquals` method is always better for unit tests.

**Comment by mguillem** [comment]
And already a new release available! Impressive. Thanks for your work.

File: src/test/java/hudson/plugins/claim/UserClaimsReportTest.java (hunk 5 above)
Comment (Greybird) [reply src/test/java/hudson/plugins/claim/UserClaimsReportTest.java:86]: The goal was to separate the actual test assertion from test setup requirement.
You are however right, would be better en ensure the issue is reported in all cases. 
Will change this.

**Comment by Greybird** [comment]
Sure!
I just wanted to have a single way throughout the code, and as Java is not my main language, and I did not check the doc, I did something wrong.
Will update using emptyXXX() methods, as this is the way to go currently.
Thanks for pointing this!

**Comment by Greybird** [comment]
Fixed everything in https://github.com/jenkinsci/claim-plugin/commit/da7f03d7d97c2c81c109fdb4b7225131e66d47ce

**Comment by mguillem** [comment]
Not bad, when Java is not your main language ;-)
